File: learn/graph/floyd-warshall/floyd_warshall.go


import "math"

type FloydWarshall struct {
	vertices int
	dist     [][]int
}

func NewFloydWarshall(v int) *FloydWarshall {
	fw := &FloydWarshall{
		vertices: v,
		dist:     make([][]int, v),
	}
	for i := 0; i < v; i++ {
		fw.dist[i] = make([]int, v)
		for j := 0; j < v; j++ {
			if i == j {
				fw.dist[i][j] = 0
			} else {
				fw.dist[i][j] = math.MaxInt32
			}
		}
	}
	return fw
}

func (fw *FloydWarshall) AddEdge(u, v, weight int) {
	fw.dist[u][v] = weight
}

func (fw *FloydWarshall) Compute() {
	v := fw.vertices
	for k := 0; k < v; k++ {
		for i := 0; i < v; i++ {
			for j := 0; j < v; j++ {
				if fw.dist[i][k] != math.MaxInt32 && fw.dist[k][j] != math.MaxInt32 && fw.dist[i][k]+fw.dist[k][j] < fw.dist[i][j] {
					fw.dist[i][j] = fw.dist[i][k] + fw.dist[k][j]
				}
			}
		}
	}
}

func (fw *FloydWarshall) GetDist() [][]int {
	return fw.dist
}

File: learn/graph/floyd-warshall/floyd_warshall.rs
pub struct FloydWarshall {
    pub vertices: usize,
    pub dist: Vec<Vec<i32>>,
}

impl FloydWarshall {
    pub fn new(vertices: usize) -> Self {
        let mut dist = vec![vec![i32::MAX; vertices]; vertices];
        for i in 0..vertices {
            dist[i][i] = 0;
        }
        FloydWarshall { vertices, dist }
    }

    pub fn add_edge(&mut self, u: usize, v: usize, weight: i32) {
        self.dist[u][v] = weight;
    }

    pub fn compute(&mut self) {
        let v_count = self.vertices;
        for k in 0..v_count {
            for i in 0..v_count {
                for j in 0..v_count {
                    if self.dist[i][k] != i32::MAX && self.dist[k][j] != i32::MAX {
                        let new_dist = self.dist[i][k] + self.dist[k][j];
                        if new_dist < self.dist[i][j] {
                            self.dist[i][j] = new_dist;
                        }
                    }
                }
            }
        }
    }
}
